aslprep/utils/confounds.py

```diff
@@ -234,72 +234,125 @@ def average_cbf_by_tissue(cbf, gm, wm, csf, thresh):
-    r1 = np.array([0, np.corrcoef(scbf[msk == 1], pbcf[msk == 1])[1, 0]]).max()
+    n_gm = np.sum(gm_mask)
+    n_wm = np.sum(wm_mask)
+    n_csf = np.sum(csf_mask)
+    # TODO: possibly add a check to ensure that n_gm, n_wm, n_csf are all greater than 1
 
     V = (
-        (np.sum(gm1) - 1) * np.var(scbf[gm1 > 0])
-        + (np.sum(wm1) - 1) * np.var(scbf[wm1 > 0])
-        + (np.sum(cc1) - 1) * np.var(scbf[cc1 > 0])
-    ) / (np.sum(gm1 > 0) + np.sum(wm1 > 0) + np.sum(cc1 > 0) - 3)
-
-    negGM = np.sum(scbf[gm1] < 0) / (np.sum(gm1))
-    GMCBF = np.mean(scbf[gm1])
-    CV = V / np.abs(GMCBF)
-    Q = [fun1(x1, CV), fun1(x2, negGM), fun2(x4, r1)]
-    return gmean(Q)
+        (n_gm - 1) * np.var(cbf_data[gm_mask])
+        + (n_wm - 1) * np.var(cbf_data[wm_mask])
+        + (n_csf - 1) * np.var(cbf_data[csf_mask])
+    ) / (n_gm + n_wm + n_csf - 3)
+
+    mean_gm_cbf = np.mean(cbf_data[gm_mask])
+    return V / np.abs(mean_gm_cbf)
+
+
+def compute_qei(
+    gm, wm, csf, img, thresh,
+    alpha=-3.0126,
+    beta=2.4419,
+    gamma=0.054,
+    delta=0.9272,
+    epsilon=2.8478,
+    zeta=0.5196,
+):
+    r"""Compute quality evaluation index (QEI) using the individual components.
 
+    The QEI is based on :footcite:t:`dolui2017automated`.
 
-def negativevoxel(cbf, gm, thresh):
-    """Compute percentage of negative voxels within grey matter mask."""
-    gm = nb.load(gm).get_fdata()
-    cbf = nb.load(cbf).get_fdata()
-    gm_bin = gm > thresh
+    .. math::
 
-    n_gm_voxels = np.sum(gm_bin)
+        QEI = \sqrt[3]{(1 - e^{\alpha\rho_{ss}^{\beta}})
+              e^{-\left(\gamma\text{DI}^{\delta} + \epsilon\text{negGM}^{\zeta}\right)}}
 
-    negative_cbf_bin = cbf < 0
-    n_negative_cbf_voxels_in_gm = np.sum(negative_cbf_bin * gm_bin)
+    Parameters
+    ----------
+    rho_ss : float
+        Structural pseudocbf correlation.
+    dispersion_index : float
+        Dispersion index (DI).
+    neg_gm : float
+        Proportion of negative voxels within grey matter mask.
 
-    return 100 * (n_negative_cbf_voxels_in_gm / n_gm_voxels)
+    Returns
+    -------
+    qei : float
+        Quality evaluation index.
+    rho_ss : float
+        Structural pseudocbf correlation.
+    di : float
+        Dispersion index (DI).
+    neg_gm : float
+        Percentage of negative voxels within grey matter mask.
+    """
+    cbf_data = _load_one_image(img)
+    smoothed_cbf = smooth_image(cbf_data, fwhm=5)
+
+    gm_probseg_data = _load_one_image(gm)
+    wm_probseg_data = _load_one_image(wm)
+    csf_probseg_data = _load_one_image(csf)
+    gm_mask = gm_probseg_data > thresh
+    wm_mask = wm_probseg_data > thresh
+    csf_mask = csf_probseg_data > thresh
+
+    # Calculate the components of the QEI equation
+    di = dispersion_index(wm_mask, gm_mask, csf_mask, smoothed_cbf, 0)
+    neg_gm = np.sum(smoothed_cbf[gm_mask] < 0) / (np.sum(gm_mask))
+    rho_ss = structural_pseudocbf_correlation(gm_probseg_data, wm_probseg_data, smoothed_cbf)
```

Comment:
I think this is going to smooth twice because it's being passed smoothed_cbf, needs smoothing_fwhm=0

Reply:
I removed the smoothing parts of dispersion_index() and structural_pseudocbf_correlation() and added documentation to prevent similar mistakes in the future